buildmimic/bigquery/README.md

```diff
@@ -0,0 +1,221 @@
+
+---
+## STEP 6: Create tables and load the csv files
+
+### A) Create a script file (ex: upload_MIMIC3_v1_4.sh) and copy the code below. Change the **schema_local_folder** to match the path to the schemas on your local machine.
+
+```
+#!/bin/bash
+
+# Initialize parameters
+bucket="mimic3_v1_4"
+dataset="MIMIC3_V1_4"
+schema_local_folder="/home/user/mimic3_schema"
+
+# Get the list of files in the bucket
+FILES=$(gsutil ls gs://$bucket)
+
+for file in $FILES
+do
+
+# Extract the table name from the file path (ex: gs://mimic3_v1_4/ADMISSIONS.csv)
+base=${file##*/}       # remove path
+tablename=${base%.*}   # remove extension .csv
+
+# Create table and populate it with data from the bucket
+bq load --allow_quoted_newlines --skip_leading_rows=1 --source_format=CSV $dataset.$tablename gs://$bucket/$tablename.csv $schema_local_folder/$tablename.schema.json
+
+# Check for error
+if [ $? -eq 0 ];then
+    echo "OK....$tablename"
+else 
+    echo "FAIL..$tablename"
+fi
+
+done
+exit 0
+```
+
+This code will get the list of files in the bucket, and for each file, it will extract the table name (ex: ADMISSIONS). With the table name, the system executes the BigQuery load command `bq load` to create a table and load the data from the csv file from the bucket using the specific schema from the local folder. The script will need the name of the bucket where the csv files are stored on Cloud Storage, the name of the dataset to create and upload the tables, and the path to the JSON file (schema) on your local machine.
+
+### B) Set the CHMOD to allow the file as executable (ex: 755), and execute the script file.
+
+```
+./upload_MIMIC3_v1_4.sh
+```
+
+During the load with the use of schemas the following error occurred: 
+- missing close double quote (") character (NOTEEVENTS)
+
+This error is related to the source CSV file containing newlines within the string (ex: column TEXT). To fix it, we added the parameter `--allow_quoted_newlines` to the command `bq load`.
+
+### C) Results of the upload process. 
+
+The system will print **OK** or **FAIL** for each file processed. In case of failure, the system also prints a message with information about the error.
+
+#### Upload OK
+
+```
+Waiting on bqjob_r7fe9abf4651b3ff3_000001620e9d13ff_1 ... (4s) Current status: DONE  
+OK....ADMISSIONS
+```
+
+#### Upload FAIL
+
+```
+Waiting on bqjob_r3c23bb4d717cd8a9_000001620e9d5f6d_1 ... (496s) Current status: DONE   
+BigQuery error in load operation: Error processing job
+'sandbox-nlp:bqjob_r3c23bb4d717cd8a9_000001620e9d5f6d_1': Error while reading data, error message: CSV table encountered too many errors, giving up. Rows: 63349; errors: 1. Please look into the error stream for more details.
+Failure details:
+- gs://mimic3_v1_4/CHARTEVENTS.csv: Error while reading data,
+error message: Could not parse 'No' as double for field VALUE
+(position 8) starting at location 3353598526
+FAIL..CHARTEVENTS
+```
```

Comment:
Consider adding a note at the end to delete the GCS bucket if it's no longer required (gsutil rm gs://mimic3_v1_4/*.csv && gsutil rb gs://mimic3_v1_4). No point paying for storage indefinitely, and safer to clean up sensitive data.